File: Taboola JS ObjC Sample/Taboola JS ObjC Sample/View controllers/FetchOnDemandPageViewController/FetchOnDemandPageViewController.m
//
//  FetchOnDemandPageViewController.m
//  Taboola JS ObjC Sample
//
//  Created by Roman Slyepko on 1/23/19.
//  Copyright © 2019 Taboola LTD. All rights reserved.
//

#import "FetchOnDemandPageViewController.h"
#import "PageContentViewController.h"

@interface FetchOnDemandPageViewController () <UIPageViewControllerDataSource>

@end

@implementation FetchOnDemandPageViewController {
    NSUInteger pageQnty;
    NSMutableArray *viewControllers;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    pageQnty = 5;
    self.dataSource = self;
    viewControllers = [NSMutableArray array];
    for (int i = 0; i < pageQnty; i++) {
        UIViewController *vc = [self viewControllerAtIndex:i];
        CGRect dontCareRect = vc.view.frame;
        [viewControllers addObject:vc];
    }
    [self setViewControllers:@[viewControllers[0]] direction:UIPageViewControllerNavigationDirectionForward animated:YES completion:nil];
    // Do any additional setup after loading the view.
}

- (PageContentViewController *)viewControllerAtIndex:(NSUInteger)index {
    if (index < 0 || index >= pageQnty)
        return nil;
    
    // Create a new view controller and pass suitable data.
    PageContentViewController *pageContentViewController = [self.storyboard instantiateViewControllerWithIdentifier:@"PageContentViewController"];
    pageContentViewController.pageIndex = index;
    
    return pageContentViewController;
}

#pragma mark - Navigation

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerAfterViewController:(UIViewController *)viewController {
    NSUInteger index = ((PageContentViewController*) viewController).pageIndex + 1;
    if (pageQnty > index) {
        return viewControllers[index];
    }
    return nil;
}

- (UIViewController *)pageViewController:(UIPageViewController *)pageViewController viewControllerBeforeViewController:(UIViewController *)viewController {
    NSUInteger index = ((PageContentViewController*) viewController).pageIndex - 1;
    if (pageQnty > index) {
        return viewControllers[index];
    }
    return nil;
}

- (NSInteger)presentationCountForPageViewController:(UIPageViewController *)pageViewController {
    return pageQnty;
}

@end

File: Taboola_SDK_Swift_Sample/Taboola SDK Swift Sample/Taboola SDK Swift Sample/SDK JS/Fetch on demand/FetchOnDemandPageViewController.swift
//
//  FetchOnDemandPageViewController.swift
//  Taboola JS Swift Sample
//
//  Created by Roman Slyepko on 2/12/19.
//  Copyright © 2019 Taboola LTD. All rights reserved.
//

import UIKit

class FetchOnDemandPageViewController: UIPageViewController {
    let pageQnty = 5
    var pageViewControllers = [UIViewController]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        dataSource = self
        if let firstViewController = viewController(at: 0) {
            self.setViewControllers([firstViewController], direction: .forward, animated: true, completion: nil)
        }
        for i in 0..<pageQnty {
            guard let vc = viewController(at: i) else {
                print("Failed to populate view controllers")
                break
            }
            _ = vc.view.frame
            pageViewControllers.append(vc)
        }
    }
    
    func viewController(at index: Int) -> PageContentViewController? {
        guard index >= 0, index < pageQnty else {
            return nil 
        }
        if pageViewControllers.count > index, let existingViewController = pageViewControllers[index] as? PageContentViewController {
            return existingViewController
        }
        let pageContentViewController = storyboard?.instantiateViewController(withIdentifier: "PageContentViewController") as? PageContentViewController
        pageContentViewController?.pageIndex = index
        return pageContentViewController
    }
}

extension FetchOnDemandPageViewController: UIPageViewControllerDataSource {
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
        guard let contentViewController = viewController as? PageContentViewController, let index = contentViewController.pageIndex else {
            return nil
        }
        return self.viewController(at: index-1)
    }
    
    func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
        guard let contentViewController = viewController as? PageContentViewController, let index = contentViewController.pageIndex else {
            return nil
        }
        return self.viewController(at: index+1)
    }
    
    func presentationCount(for pageViewController: UIPageViewController) -> Int {
        return pageQnty
    }
}
